geomstats/geometry/poincare_ball.py

```diff
@@ -317,3 +317,268 @@ def inner_product_matrix(self, base_point=None):
+            self.normalization_factor(variances, dimension)
+
+        cond_1 = self.normalisation_factor_var.sum() != \
+            self.normalisation_factor_var.sum()
+        cond_2 = self.normalisation_factor_var.sum() == float('+inf')
+        cond_3 = self.normalisation_factor_var.sum() == float('-inf')
+
+        if cond_1 or cond_2 or cond_3:
+            print('WARNING :\n'
+                  'untracktable normalisation factor :')
+            max_nf = len(variances)
+
+            limit_nf = ((self.normalisation_factor_var /
+                         self.normalisation_factor_var)
+                        * 0).nonzero()[0].item()
+            self.variances = self.variances[0:limit_nf]
+            self.normalisation_factor_var = \
+                self.normalisation_factor_var[0:limit_nf]
+            if cond_1:
+                print('\t Nan value in processing normalisation factor')
+            if cond_2 or cond_3:
+                print('\t +-inf value in processing normalisation factor')
+
+            print('\t Max variance is now : ', self.variances[-1])
+            print('\t Number of possible variance is now: '
+                  + str(len(self.variances)) + '/' + str(max_nf))
+
+        _, log_grad_zeta = \
+            self.norm_factor_gradient(self.variances, dimension)
+
+        self.phi_inv_var = self.variances ** 3 * log_grad_zeta
+
+    def find_normalisation_factor(self, variance):
+        """Given a variance, finds the normalisation factor."""
+        N, P = variance.shape[0], self.variances.shape[0]
+
+        ref = gs.expand_dims(self.variances, 0)
+        ref = gs.repeat(ref, N, axis=0)
+        val = gs.expand_dims(variance, 1)
+        val = gs.repeat(val, P, axis=1)
+
+        difference = gs.abs(ref - val)
+
+        index = gs.argmin(difference, axis=-1)
+
+        return self.normalisation_factor_var[index]
+
+    def find_variance_from_index(self, weighted_distances):
+        """Return the variance given weighted distances."""
+        n_gaussian, precision = \
+            weighted_distances.shape[0], self.variances.shape[0]
+
+        ref = gs.expand_dims(self.phi_inv_var, 0)
+        ref = gs.repeat(ref, n_gaussian, axis=0)
+
+        val = gs.expand_dims(weighted_distances, 1)
+        val = gs.repeat(val, precision, axis=1)
+
+        abs_difference = gs.abs(ref - val)
+
+        index = gs.argmin(abs_difference, -1)
+
+        return self.variances[index]
+
+    @classmethod
+    def normalization_factor(cls, variances, dimension):
+        """Return normalization factor."""
+        binomial_coefficient = None
+        n_samples = variances.shape[0]
+
+        expand_variances = gs.expand_dims(variances, axis=0)
+        expand_variances = gs.repeat(expand_variances, dimension, axis=0)
+
+        if binomial_coefficient is None:
+
+            v = gs.arange(dimension)
+            v[0] = 1
+            n_fact = v.prod()
+
+            k_fact = gs.concatenate([gs.expand_dims(v[:i].prod(), 0)
```

Comment:
NIT: I usually prefer the following indentation:
```
my_var = ma_function(
    arg1, arg2)
```
rather than
```
my_var = ma_function(arg1, 
                     arg2)
```
because the first one is robust if we make the change my_var -> my_favorite_var (doesn't break flake8), while the second gets an indentation problem.